### PR TITLE
[15.0][IMP][account_financial_report]: add custom intervals for aged partner report

### PR DESCRIPTION
**Pull request by ernesto-garcia-tecnativa**
@tecnativa TT45574

### CODE REVIEW

**Comment by carolinafernandez-tecnativa** [comment]
WIP

**Comment by carolinafernandez-tecnativa** [comment]
@pedrobaeza could you please check it now? Thanks

File: account_financial_report/models/account_age_report_configuration.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class AccountAgeReportConfigurationLine(models.Model):
+    _name = "account.age.report.configuration.line"
+    _description = "Model to set interval lines for Age partner balance report"
+
+    name = fields.Char(required=True)
+    code = fields.Char(required=True)
+    account_age_report_config_id = fields.Many2one("account.age.report.configuration")
+    lower_limit = fields.Integer()
```
Comment (pedrobaeza) [inline account_financial_report/models/account_age_report_configuration.py]: We talked here about putting only one value, so no need to check overlappings and missing gaps:

Configuration 1:
- 15
- 30
- 60

It means the first interval is from 0 to 15, the second from 16 to 30, and the third is  61+.

File: account_financial_report/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Tecnativa - Carolina Fernandez
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    age_partner_config_id = fields.Many2one(
```
Comment (pedrobaeza) [inline account_financial_report/models/res_company.py]: Don't store this in company.

File: account_financial_report/models/res_config_settings.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    default_age_partner_config_id = fields.Many2one(
+        "account.age.report.configuration",
+        related="company_id.age_partner_config_id",
```
Comment (pedrobaeza) [inline account_financial_report/models/res_config_settings.py]: Don't mix related and `default_*`. A default is for all companies.

File: account_financial_report/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -14,3 +14,5 @@ currency used in account move lines is properly shown.
 
 In case that in an account has not been configured a second currency foreign
 currency balances are not available.
+
+Invoicing / Settings / Invoicing / OCA Aged Report Configuration you will be able to set dynamic intervals that will appear on the Aged Partner Balance
```
Comment (pedrobaeza) [inline account_financial_report/readme/DESCRIPTION.rst]: This should go better redacted on CONFIGURE.rst. And please respect 80 cols limit.

File: account_financial_report/report/aged_partner_balance_xlsx.py [hunk 5]
```diff
@@ -20,67 +21,84 @@ def _get_report_name(self, report, data=False):
             report_name = report_name + suffix
         return report_name
 
-    def _get_report_columns(self, report):
-        if not report.show_move_line_details:
-            return {
-                0: {"header": _("Partner"), "field": "name", "width": 70},
-                1: {
-                    "header": _("Residual"),
```
Comment (pedrobaeza) [inline account_financial_report/report/aged_partner_balance_xlsx.py:28]: Residual and current columns should be respected, no matter the configuration.

File: account_financial_report/report/aged_partner_balance_xlsx.py [hunk 6]
```diff
@@ -96,61 +114,84 @@ def _get_report_columns(self, report):
...
-            },
-            13: {
-                "header": _("Older"),
-                "field": "older",
-                "field_footer_total": "older",
-                "field_footer_percent": "percent_older",
-                "field_final_balance": "older",
+        }
+        if not report.age_partner_config_id:
+            report_columns.update(
```
Comment (pedrobaeza) [inline account_financial_report/report/aged_partner_balance_xlsx.py:128]: Maybe it's better to use a hidden configuration (with active=False) with all these lines, and prevent its removal than hardcode all these lines. What do you think?

File: account_financial_report/report/templates/aged_partner_balance.xml [hunk 7]
```diff
@@ -114,18 +114,24 @@
...
-                <div class="act_as_cell" style="width: 9.64%;">1 - 30 d.</div>
-                <!--## age_60_days-->
-                <div class="act_as_cell" style="width: 9.64%;">31 - 60 d.</div>
-                <!--## age_90_days-->
-                <div class="act_as_cell" style="width: 9.64%;">61 - 90 d.</div>
-                <!--## age_120_days-->
-                <div class="act_as_cell" style="width: 9.64%;">91 - 120 d.</div>
-                <!--## older-->
-                <div class="act_as_cell" style="width: 9.64%;">> 120 d.</div>
+                <t t-if="not age_partner_config">
```
Comment (pedrobaeza) [inline account_financial_report/report/templates/aged_partner_balance.xml:119]: Here, it shouldn't be needed the `if`, as it should be already got in the Python structure.

File: account_financial_report/report/templates/aged_partner_balance.xml [hunk 8]
```diff
@@ -143,48 +149,58 @@
...
-                />
-            </div>
-            <!--## older-->
-            <div class="act_as_cell amount">
-                <span
-                    t-esc="partner['older']"
-                    t-options="{'widget': 'monetary', 'display_currency': res_company.currency_id}"
-                />
-            </div>
+            <t t-if="not age_partner_config">
```
Comment (pedrobaeza) [inline account_financial_report/report/templates/aged_partner_balance.xml]: The same here: you shouldn't put an if.

File: account_financial_report/report/templates/aged_partner_balance.xml [hunk 9]
```diff
@@ -216,30 +232,38 @@
                     </div>
                     <!--## amount_residual-->
                     <div class="act_as_cell" style="width: 6.00%;">Residual</div>
+                    <t t-if="not age_partner_config">
```
Comment (pedrobaeza) [inline account_financial_report/report/templates/aged_partner_balance.xml:237]: The same...

File: account_financial_report/report/templates/aged_partner_balance.xml [hunk 10]
```diff
@@ -333,126 +357,149 @@
...
-                            />
-                        </t>
-                        <t t-else="">
-                            <span
-                                domain="[('id', 'in', (line['line_rec'] | line['line_rec'].matched_debit_ids.mapped('debit_move_id') | line['line_rec'].matched_credit_ids.mapped('credit_move_id')).ids)]"
-                                res-model="account.move.line"
-                            >
-                                <t
-                                    t-out="line['current']"
+                    <t t-if="not age_partner_config">
```
Comment (pedrobaeza) [inline account_financial_report/report/templates/aged_partner_balance.xml]: And the same...

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by carolinafernandez-tecnativa** [comment]
WIP

File: account_financial_report/report/aged_partner_balance_xlsx.py (hunk 5 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/report/aged_partner_balance_xlsx.py:28]: Done

File: account_financial_report/models/account_age_report_configuration.py (hunk 1 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/models/account_age_report_configuration.py]: Done

File: account_financial_report/models/res_config_settings.py (hunk 3 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/models/res_config_settings.py]: Done

File: account_financial_report/readme/DESCRIPTION.rst (hunk 4 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/readme/DESCRIPTION.rst]: Done

File: account_financial_report/models/res_company.py (hunk 2 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/models/res_company.py]: Done

File: account_financial_report/report/templates/aged_partner_balance.xml (hunk 7 above)
Comment (carolinafernandez-tecnativa) [reply account_financial_report/report/templates/aged_partner_balance.xml:119]: If it is not added, it will show all columns and it is not correct.

**Review by chienandalu — COMMENTED** [review]
Could you resolve conflicts @carolinafernandez-tecnativa ?

**Comment by carolinafernandez-tecnativa** [comment]
> Could you resolve conflicts @carolinafernandez-tecnativa ?

Done!

**Comment by chienandalu** [comment]
New conflicts :( When you can @carolinafernandez-tecnativa

**Comment by carolinafernandez-tecnativa** [comment]
> New conflicts :( When you can @carolinafernandez-tecnativa

Done!

**Comment by pedrobaeza** [comment]
@carolinafernandez-tecnativa several things:

- I can't delete configuration lines.
- It puts "Superior limit", but it should "Inferior limit" AFAIK, as it's the starting days from which the interval is activated.
- Why having "code" and "name"? I think the code (for internal purposes) should be automatic. It can be the `ID` for example.

**Comment by carolinafernandez-tecnativa** [comment]
> @carolinafernandez-tecnativa several things:
> 
> * I can't delete configuration lines.
> * It puts "Superior limit", but it should "Inferior limit" AFAIK, as it's the starting days from which the interval is activated.
> * Why having "code" and "name"? I think the code (for internal purposes) should be automatic. It can be the `ID` for example.

Thanks!
Fixed first 2 points, WIP with third point.

**Comment by pedrobaeza** [comment]
Tested on runboat, the limits are not working well.

Check the example in http://oca-account-financial-reporting-15-0-pr1083-c23b2616a35d.runboat.odoo-community.org with "Test customer", that has an invoice with due date 2024-03-27, and when reporting for that date, the amount appears in both the "Current" column and in "5 days" one. And you have to overpass 10 days to see the amount in the column "10 days". I think that the computation is still based on superior limit, not on inferior one.

**Comment by pedrobaeza** [comment]
It has still some edges:

- For an interval with superior limit 1 and a invoice with due today, it appears in current debt + the 1 interval:

imagen

- Totals are not taking into account:

imagen

**Comment by carolinafernandez-tecnativa** [comment]
> It has still some edges:
> 
> * For an interval with superior limit 1 and a invoice with due today, it appears in current debt + the 1 interval:
> 
> imagen
> 
> * Totals are not taking into account:
> 
> imagen

Could you please check now?

I fw theses changes to v16

**Comment by carolinafernandez-tecnativa** [comment]
> > It has still some edges:
> > 
> > * For an interval with superior limit 1 and a invoice with due today, it appears in current debt + the 1 interval:
> > 
> > imagen
> > 
> > * Totals are not taking into account:
> > 
> > imagen
> 
> Could you please check now?
> 
> I fw theses changes to v16

Hi @pilarvargas-tecnativa could you please help us to validate? thanks

**Review by pilarvargas-tecnativa — APPROVED** [review]
I have run the test in runbot and the ranges in which the overdue invoices appear in the report are correct. Thanks :+1:

**Review by pedrobaeza — APPROVED** [review]
Finally!

/ocabot merge minor

**Pull request merged**